### PR TITLE
Preserve CIA ops in AOT joint export

### LINKED ISSUE #160388: Perf differences observed between AOT and JIT on RAFT model
## Perf Issue
The latency of AOT (export + compile) and JIT (torch.compile) on RAFT model are very different. I expect AOT and JIT to be similar but based on my observation, AOT is the fastest, and then Original Pytorch model, and then JIT is the slowest.

I got the results below by running the following repro on NVIDIA GeForce RTX 4080.

Original time: 6.775218769907951 ms
AOT time: 2.546852830564603 ms
JIT time: 17.458299309946597 ms

## Repro
```
#!/usr/bin/env python3
"""
Compare the perf of RAFT-small and/or RAFT-large for original pytorch model, AOT, and JIT.
"""
import torch_tensorrt
import torch
from torchvision.models.optical_flow import raft_small, raft_large
import timeit

WARM_UP_ITERS = 10
NUM_ITERS = 100


def benchmark(model, inputs):
    # Warmup
    for _ in range(WARM_UP_ITERS):
        with torch.no_grad():
            _ = model(inputs[0], inputs[1])
    
    # Benchmark
    torch.cuda.synchronize()
    start = timeit.default_timer()
    for _ in range(NUM_ITERS):
        with torch.no_grad():
            _ = model(inputs[0], inputs[1])
    torch.cuda.synchronize()
    inference_time = (timeit.default_timer() - start) / NUM_ITERS
    return inference_time


def analyze_raft_operations(model_type='small', target_size=(240, 424)):
    # Load model
    if model_type == 'small':
        model = raft_small(pretrained=True, progress=False)
    else:
        model = raft_large(pretrained=True, progress=False)
    
    model = model.cuda().eval()

    # Adjust dimensions
    height = (target_size[0] // 8) * 8
    width = (target_size[1] // 8) * 8
    
    inputs = [
        torch.randn(1, 3, height, width).cuda(),
        torch.randn(1, 3, height, width).cuda()
    ]
    
    print(f"Model: RAFT-{model_type}")
    print(f"Input size: {height}x{width}")
    
    print("--------------Original--------------")
    original_time = benchmark(model, inputs)
    
    print("--------------AOT--------------")
    exported_program = torch.export.export(model, tuple(inputs))
    aot_compiled_model = torch_tensorrt.dynamo.compile(
        exported_program, 
        tuple(inputs), 
        use_python_runtime=True,
    )
    aot_time = benchmark(aot_compiled_model, inputs)
    
    print("--------------JIT--------------")
    with torch.no_grad():
        jit_compiled_model = torch.compile(
            model,
            backend="tensorrt",
            options={
                "use_python_runtime": True,
            }
        )
    jit_time = benchmark(jit_compiled_model, inputs)
    
    print(f"Original time: {original_time * 1000} ms")
    print(f"AOT time: {aot_time * 1000} ms")
    print(f"JIT time: {jit_time * 1000} ms")


if __name__ == "__main__":
    analyze_raft_operations(model_type='small')
```

### Error logs

No error, just perf issue

### Versions

Collecting environment information...
PyTorch version: 2.9.0.dev20250804+cu129
Is debug build: False
CUDA used to build PyTorch: 12.9
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.13.2 | packaged by Anaconda, Inc. | (main, Feb  6 2025, 18:56:02) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-49-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.9.86
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4080
Nvidia driver version: 570.169
cuDNN version: Probably one of the following:
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn.so.8.9.3
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.9.3
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.9.3
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.9.3
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.9.3
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.9.3
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.9.3
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            GenuineIntel
Model name:                           13th Gen Intel(R) Core(TM) i9-13900KF
CPU family:                           6
Model:                                183
Thread(s) per core:                   2
Core(s) per socket:                   24
Socket(s):                            1
Stepping:                             1
CPU max MHz:                          6000.0000
CPU min MHz:                          800.0000
BogoMIPS:                             5990.40
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            896 KiB (24 instances)
L1i cache:                            1.3 MiB (24 instances)
L2 cache:                             32 MiB (12 instances)
L3 cache:                             36 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Vulnerable: No microcode
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] mypy==1.15.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.3.1
[pip3] nvidia-cublas-cu12==12.9.1.4
[pip3] nvidia-cuda-cupti-cu12==12.9.79
[pip3] nvidia-cuda-nvrtc-cu12==12.9.86
[pip3] nvidia-cuda-runtime-cu12==12.9.79
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.4.1.4
[pip3] nvidia-curand-cu12==10.3.10.19
[pip3] nvidia-cusolver-cu12==11.7.5.82
[pip3] nvidia-cusparse-cu12==12.5.10.65
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.9.86
[pip3] nvidia-nvtx-cu12==12.9.79
[pip3] onnx==1.18.0
[pip3] onnx-ir==0.1.4
[pip3] onnxscript==0.3.1
[pip3] pytorch_sphinx_theme==0.0.24
[pip3] pytorch-triton==3.4.0+git11ec6354
[pip3] torch==2.9.0.dev20250804+cu129
[pip3] torch_tensorrt==2.9.0.dev0+df8a23897
[pip3] torchaudio==2.8.0.dev20250804+cu129
[pip3] torchvision==0.24.0.dev20250804+cu129
[pip3] triton==3.2.0
[conda] numpy                     2.3.1                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.9.1.4                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.9.79                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.9.86                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.9.79                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.10.2.21                pypi_0    pypi
[conda] nvidia-cufft-cu12         11.4.1.4                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.10.19               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.5.82                pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.10.65               pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.7.1                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.27.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.9.86                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.9.79                  pypi_0    pypi
[conda] pytorch-triton            3.4.0+git11ec6354          pypi_0    pypi
[conda] torch                     2.9.0.dev20250804+cu129          pypi_0    pypi
[conda] torch-tensorrt            2.9.0.dev0+df8a23897          pypi_0    pypi
[conda] torchaudio                2.8.0.dev20250804+cu129          pypi_0    pypi
[conda] torchvision               0.24.0.dev20250804+cu129          pypi_0    pypi
[conda] triton                    3.2.0                    pypi_0    pypi

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184052

Apply export's CompositeImplicitAutograd decomposition table handling to aot_export_joint_simple so backend AOT graphs agree with export on preserved functional CIA ops. Keep direct aot_export_module decomposition behavior unchanged.

Fixes #160388

Generated by my agent

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by ezyang — DISMISSED** [review]
change has code motion which makes it difficult for human to review in github ui, please amend agent code authoring instructions

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -1528,6 +1528,32 @@ def unflattened_compiled_fn(*args: Any, **kwargs: Any) -> Any:
     return unflattened_compiled_fn
 
 
+@contextlib.contextmanager
+def _aot_export_decomposition_context(
+    decompositions: dict[OpOverload, Callable[..., Any]] | None,
+) -> Any:
```
Comment (aorenste) [inline torch/_functorch/aot_autograd.py]: Can we get a better return type than `Any`?

**Review by aorenste — APPROVED** [review]
Needs a test for the decompositions=None behavior...

Approving w/ comments

**Comment by jansel** [comment]
@claude review these changes

File: torch/_functorch/aot_autograd.py (hunk 1 above)
Comment (jansel) [reply torch/_functorch/aot_autograd.py]: My agent says this is addressed in the latest revision: `_aot_export_decomposition_context` now returns `Iterator[dict[OpOverload, Callable[..., Any]] | None]` instead of `Any`.

**Comment by jansel** [comment]
My agent says the code-motion review feedback is addressed in the latest revision: the net diff now only changes `torch/_functorch/aot_autograd.py` and `test/functorch/test_aotdispatch.py`, leaving the export helper move out of the PR.

**Comment by jansel** [comment]
@ezyang can you review again?  Your prior review will block merging.

**Comment by ezyang** [comment]
deferring to the actual reviewer

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -1780,14 +1806,15 @@ def aot_export_joint_simple(
...
 
-    with ctx():
-        fx_g, metadata, in_spec, out_spec = _aot_export_function(
-            func,
-            args,
-            decompositions=decompositions,
-            trace_joint=trace_joint,
-        )
-        in_spec, _kw_in_spec = in_spec.children()
+    with _aot_export_decomposition_context(decompositions) as decompositions_for_aot:
```
Comment (mlazos) [inline torch/_functorch/aot_autograd.py:1809]: _Reviewed by me; this finding applies here._

_AI-generated analysis (Codex):_

> **CIA override drops gradients from joint graphs**
> 
> The new context also wraps trace_joint=True. With decompositions={}, preserved aten.matmul receives a deferred autograd-not-implemented handler, disconnecting its result from both inputs; joint capture consequently emits (None, None) gradients. The added joint test checks only the forward target and misses the unusable backward.

File: torch/_functorch/aot_autograd.py [hunk 3]
```diff
@@ -1528,6 +1528,32 @@ def unflattened_compiled_fn(*args: Any, **kwargs: Any) -> Any:
...
+
+    decomp_table = (
+        decompositions.materialize()
+        if isinstance(decompositions, CustomDecompTable)
+        else cast("dict[OperatorBase, Callable[..., Any]]", dict(decompositions))
+    )
+    cia_to_decomp, python_decomp_table = _split_decomp_table_to_cia_and_python_decomp(
+        decomp_table
+    )
+    with _override_composite_implicit_decomp(cia_to_decomp):
```
Comment (mlazos) [inline torch/_functorch/aot_autograd.py:1553]: _Reviewed by me; this finding applies here._

_AI-generated analysis (Codex):_

> **Context removes pre-existing FakeTensor handlers**
> 
> For every non-None table, the split includes all preservable CIA ops. The reused override skips fake-rule installation for already-registered ops but unconditionally deregisters every patched op on exit, removing the process-global handlers for aten.item.default, aten.nonzero_numpy.default, and aten._convolution.deprecated. Later FakeTensor dispatch no longer uses those dedicated implementations, and nested contexts can tear down an outer context's registrations.
